stl/inc/span

```
@@ -270,6 +270,19 @@ concept _Span_compatible_range =
     && _RANGES contiguous_range<_Rng> && _RANGES sized_range<_Rng> && (_RANGES borrowed_range<_Rng> || is_const_v<_Ty>)
     && is_convertible_v<remove_reference_t<_RANGES range_reference_t<_Rng>> (*)[], _Ty (*)[]>;
 
+template <class _Ty>
+concept _Integral_constant_like =
+    is_integral_v<remove_cvref_t<decltype(_Ty::value)>> && !is_same_v<bool, remove_const_t<decltype(_Ty::value)>>
+    && convertible_to<_Ty, decltype(_Ty::value)> && equality_comparable_with<_Ty, decltype(_Ty::value)>
+    && bool_constant<_Ty() == _Ty::value>::value
+    && bool_constant<static_cast<decltype(_Ty::value)>(_Ty()) == _Ty::value>::value;
+
+template <class _Ty>
+constexpr size_t _Maybe_static_ext = dynamic_extent;
+
+template <_Integral_constant_like _Ty>
+constexpr size_t _Maybe_static_ext<_Ty> = {_Ty::value};

```

No change requested. An LWG issue seems needed to me.

I'm somehow sure that this can result in _hard error_ due to narrowing conversion. E.g. if `_Ty::value` is of an extended integral type (not possible for MSVC STL, though), or if `_Ty::value` is 64-bit while `size_t` is 32-bit. But libc++ and libstdc++ seem to suppress the error in system headers.

I think we should still reject narrowing conversion, but not here. CC @hewillk.

---

Hum, does it matter? It's exposition-only, which means it doesn't actually exist. Users can't explicitly spell it.

---

> Hum, does it matter? It's exposition-only, which means it doesn't actually exist. Users can't explicitly spell it.

This or some equivalent thing needs to be touched by the updated deduction guides. When computing the deduced type of `span(p, integral_constant<uint64_t, N>{})` (assuming 32-bit platform), there will be an error due to narrowing conversion which not in the immediate context.

---

Here's an example ([Godbolt link](https://godbolt.org/z/5768jx9Ms)):

```C++
#include <span>
#include <type_traits>

template<class X, class Y>
concept can_ctad_span_nonlist = requires(X&& x, Y&& y) {
  std::span(static_cast<X&&>(x), static_cast<Y&&>(y));
};
template<class X, class Y>
concept can_ctad_span_list = requires(X&& x, Y&& y) {
  std::span{static_cast<X&&>(x), static_cast<Y&&>(y)};
};

static_assert(can_ctad_span_nonlist<int*, std::integral_constant<long long, 1>>, "good");
#ifdef __clang__
static_assert(can_ctad_span_nonlist<int*, std::integral_constant<long long, -1>>, "should be rejected");
#else
static_assert(can_ctad_span_nonlist<int*, std::integral_constant<long long, -1>> && true, "???");
#endif
static_assert(can_ctad_span_list<int*, std::integral_constant<long long, 1>>, "good");
static_assert(!can_ctad_span_list<int*, std::integral_constant<long long, -1>>, "good");
```

Currently, GCC conformingly emits hard error for this (due to the seemingly exposition-only, but actually normative list-initialization), while Clang incorrectly suppresses the error. But it's quite weird to give a hard error for just asking whether CTAD is valid or not.